### PR TITLE
[SDK-3581] Add v2 migration guide

### PR DESCRIPTION
**Pull request by frederikprijck**
### Changes

This PR adds the migration guide to help migrate users from v1 to v2.
This does not yet incorporate the guide in the readme, which will be done when we update the readme for v2 beta.

### Checklist

- [x] I have read the [Auth0 general contribution guidelines](https://github.com/auth0/open-source-template/blob/master/GENERAL-CONTRIBUTING.md)
- [x] I have read the [Auth0 Code of Conduct](https://github.com/auth0/open-source-template/blob/master/CODE-OF-CONDUCT.md)
- [x] All code quality tools/guidelines have been run/followed

### CODE REVIEW

File: MIGRATION_GUIDE.md [hunk 1]
```diff
@@ -0,0 +1,344 @@
...
+    - [defaultScope](#defaultscope)
+  - [checkSession no longer throws exceptions](#checksession-no-longer-throws-exceptions)
+  - [getIdTokenClaimsOptions type has been removed](#getidtokenclaimsoptions-type-has-been-removed)
+- [Module output](#module-output)
+  - [No more default export](#no-more-default-export)
+  - [Changes on how to create an instance when relying on globally available API’s (e.g. using CDN)](#changes-on-how-to-create-an-instance-when-relying-on-globally-available-apis-eg-using-cdn)
+
+## Polyfills and supported browsers
+
+As [Microsoft has dropped support for IE11](https://blogs.windows.com/windowsexperience/2022/06/15/internet-explorer-11-has-retired-and-is-officially-out-of-support-what-you-need-to-know) themselves, as of v2 our SDK <u>no longer supports IE11</u>. Because of that, we have dropped including any polyfill in our bundle, as all of these polyfills were for IE11. On top of that, our bundle output is now set to ES2017, which should work fine on all major browsers.
```
Comment (stevehobbsdev) [inline MIGRATION_GUIDE.md:29]: I would reword bits of this:

```suggestion
As [Microsoft has dropped support for IE11](https://blogs.windows.com/windowsexperience/2022/06/15/internet-explorer-11-has-retired-and-is-officially-out-of-support-what-you-need-to-know), we are no longer including any polyfills in our bundle, as all of these polyfills were for IE11. Therefore <u>we no longer support IE11 from v2 of this SDK</u>. In addition, our bundle output is now set to ES2017, which should work fine on all major browsers.
```

Comment (stevehobbsdev) [reply MIGRATION_GUIDE.md:29]: Additionally, will we provide a list anywhere of the polyfills that should be used if the app developer wants to include them themselves?

File: MIGRATION_GUIDE.md [hunk 2]
```diff
@@ -0,0 +1,344 @@
...
+
+Because of this, we have <u>dropped 60% in bundle size</u>, ensuring your users have a better experience when integrating Auth0 using our SPA-JS SDK.
+
+## Public API changes
+
+With the release of this new major version, a couple of changes were made that affect the public API of the Auth0-SPA-JS SDK. Most of these should be noticed by TypeScript, however, it’s advised to take the time to go through this list thoroughly.
```
Comment (stevehobbsdev) [inline MIGRATION_GUIDE.md:35]:
```suggestion
With the release of this new major version, a couple of changes were made that affect the public API of the Auth0-SPA-JS SDK. Most of these should be noticed by TypeScript. However, it’s advised to take the time to go through this list thoroughly.
```

File: MIGRATION_GUIDE.md [hunk 3]
```diff
@@ -0,0 +1,344 @@
...
+
+```ts
+const client = new Auth0Client({ clientId: '' });
+```
+
+This change needs to occur with every method on `Auth0Client` that takes a client id.
+
+### Introduction of `authorizationParams`
+
+Another breaking change that will affect pretty much everyone is the introduction of `authorizationParams`, a more structured approach to provide (additional) parameters to Auth0.
```
Comment (stevehobbsdev) [inline MIGRATION_GUIDE.md:55]:
```suggestion
Another breaking change that will affect pretty much everyone is the introduction of `authorizationParams`, a more structured approach to providing parameters - including custom parameters - to Auth0.
```

File: MIGRATION_GUIDE.md [hunk 4]
```diff
@@ -0,0 +1,344 @@
...
+await client.loginWithRedirect({
+  appState: {
+    key: value // state to restore when getting redirected back
+  }
+  screen_hint: 'signup', // 1st-class property to send to Auth0
+  any_custom_property: 'value' // Any additional custom property to send to Auth0
+});
+```
+
+With v2 of our SDK, we have improved the API by separating properties to configure the SDK from properties that are only used to send to Auth0. The SDK configuration properties will stay on the root, while any property that should be sent to Auth0 is expected to be set on `authorizationParams`.
```
Comment (stevehobbsdev) [inline MIGRATION_GUIDE.md:75]: Conciseness:

```suggestion
With v2 of our SDK, we have improved the API by separating those properties used to configure the SDK, from properties that are sent to Auth0. The SDK configuration properties will stay on the root, while any property that should be sent to Auth0 should be set on `authorizationParams`.
```

File: MIGRATION_GUIDE.md [hunk 5]
```diff
@@ -0,0 +1,344 @@
...
+
+```ts
+await client.loginWithRedirect({
+  onRedirect(url) {
+    window.location.replace(url);
+  }
+});
+```
+
+### `ignoreCache` on `getTokenSilentlyhas` been replaced by `cacheMode`
```
Comment (stevehobbsdev) [inline MIGRATION_GUIDE.md:193]:
```suggestion
### `ignoreCache` on `getTokenSilently` has been replaced by `cacheMode`
```

File: MIGRATION_GUIDE.md [hunk 6]
```diff
@@ -0,0 +1,344 @@
...
+- **off**: ignore the cache, instead always call Auth0
+- **cache-only**: read from the cache, don’t fall back to Auth0
+
+Any code that was previously using `ignoreCache: true` should be changed to use `cacheMode: 'off'`:
+
+```ts
+const token = await client.getTokenSilently({ cacheMode: 'off' });
+```
+
+### `application/x-www-form-urlencoded` is used by default instead of `application/json`
```
Comment (stevehobbsdev) [inline MIGRATION_GUIDE.md:213]: Might be worth mentioning in here somewhere that this provides a small performance benefit.

File: MIGRATION_GUIDE.md [hunk 7]
```diff
@@ -0,0 +1,344 @@
...
+With v2, we have flipped the default value for `useFormData` to **true**, meaning we will be sending requests to Auth0’s token endpoint using `application/x-www-form-urlencoded` as the content type by default.
+
+> :warning: This can affect existing rules and actions, and it’s important to ensure all your actions still work as expected after upgrading to v2.
+> To restore the original behaviour, you can set `useFormData` to **false**, and your rules and actions should continue to work as before.
+
+### No more iframe fallback by default when using refresh tokens
+
+When using refresh tokens in v1, we fall back to using iframes whenever a refresh token exchange would fail. This has caused problems before in environments that do not support iframes, and we have specifically introduced `useRefreshTokensFallback` to be able to opt-out of falling back to iframes in the case a refresh_grant fails.
+
+With v2, we have flipped the default value for `useRefreshTokensFallback` to false, meaning, by default, we will not be falling back to using iframes when `useRefreshTokens` is set to true and exchanging the refresh token fails.
```
Comment (stevehobbsdev) [inline MIGRATION_GUIDE.md:228]: Conciseness:

```suggestion
With v2, we have flipped the default value for `useRefreshTokensFallback` to false we do not fall back to using iframes by default when `useRefreshTokens` is `true`, and the refresh token exchange fails.
```

File: MIGRATION_GUIDE.md [hunk 8]
```diff
@@ -0,0 +1,344 @@
...
+
+If you would want to restore the original behaviour, and still fallback to iframes when the refresh_grant fails, you can set `useRefreshTokensFallback` to true.
```
Comment (stevehobbsdev) [inline MIGRATION_GUIDE.md:230]:
```suggestion
If you want to restore the original behaviour, and still fall back to iframes when the refresh token exchange fails, you can set `useRefreshTokensFallback` to true.
```

File: MIGRATION_GUIDE.md [hunk 9]
```diff
@@ -0,0 +1,344 @@
...
+
+### `getUser` and `getIdTokenClaims`
+
+With v1 of our SDK, both `getUser` and `getIdTokenClaims` support providing an (optional) audience and scope to retrieve the user information.
```
Comment (stevehobbsdev) [inline MIGRATION_GUIDE.md:234]:
```suggestion
With v1 of our SDK, both `getUser` and `getIdTokenClaims` supported optional audience and scope parameters when retrieving the user profile.
```

File: MIGRATION_GUIDE.md [hunk 10]
```diff
@@ -0,0 +1,344 @@
...
+
+This means that the following code in v1:
+
+```ts
+const client = new Auth0Client({
+  scope: 'scope1'
+});
+```
+
+Needs to be updated to explicitly include profile email to achieve the same in v2:
```
Comment (stevehobbsdev) [inline MIGRATION_GUIDE.md:267]:
```suggestion
Needs to be updated to explicitly include the `profile email` scopes to achieve the same in v2:
```

File: MIGRATION_GUIDE.md [hunk 11]
```diff
@@ -0,0 +1,344 @@
...
+
+```ts
+const client = new Auth0Client({
+  scope: 'profile email scope1'
+});
+```
+
+#### defaultScope
```
Comment (stevehobbsdev) [inline MIGRATION_GUIDE.md:275]: I would explicitly clarify here as well that `advancedOptions` no longer exists (if that's the case, I'm assuming it is). People might not realise that it only had one option in it.

File: MIGRATION_GUIDE.md [hunk 12]
```diff
@@ -0,0 +1,344 @@
...
+const client = createAuth0Client();
+const client = new Auth0Client();
+const errorTypes = [GenericError, ...]
+```
+
+The above syntax is, for the most part, also available in v1, so this is not something entirely new. The thing that’s changed is the fact that `createAuth0Client` is now also a named export, without any additional methods attached to it.
+
+### Changes on how to create an instance when relying on globally available API’s (e.g. using CDN)
+
+On top of that, we also used to have a globally available createAuth0Client and Auth0Client, which is typically used for applications that do not use a module loader and/or bundler.
```
Comment (stevehobbsdev) [inline MIGRATION_GUIDE.md:337]:
```suggestion
On top of that, we also used to have a globally available `createAuth0Client` and `Auth0Client`, which is typically used for applications that do not use a module loader and/or bundler.
```

**Review by stevehobbsdev — CHANGES_REQUESTED** [review]
LGMT, left a few suggestions 👍🏻

File: MIGRATION_GUIDE.md (hunk 1 above)
Comment (frederikprijck) [reply MIGRATION_GUIDE.md]: We have elaborated on that in the PR description, which will be linked from the changelog: https://github.com/auth0/auth0-spa-js/pull/951. 

Maybe it makes sense to add every PR to the corresponding section in the migration guide, but I don't want to cause to much confusion by pulling people to PRs if they don't need to.

**Review by stevehobbsdev — APPROVED** [review]
